### ISSUE TITLE
[Question][FR] ComponentDefininition#save_as return value

### ISSUE THREAD

**Issue by Eneroth3** [SketchUp/api-issue-tracker#249]
As many other API methods, ComponentDefinition#save_as is documented as returning true on success. However, it appears to return true even when the file wasn't written to, e.g. when the file was opened in a separate SketchUp instance or when the user lacks writing privilege.

I've made a wrapping method that checks the file modification time both before and after save_as to see if it was written to, and otherwise raise an error.

However I'm curious if the return value of save_as is ever false, and if so, under what conditions.

In addition to this question, I'd be happy if save_as could return different IOErrors based on what caused the save to fail, but maybe that wouldn't be compatible with existing extensions that expects a false return value.

**Comment by DanRathbun** [comment]
Raising a proper `IOError` (or an `Errno` subclass) is so mush more robust than just returning `false`. When `false` is returned, the code is likely wanting to respond in certain ways depending upon what went wrong. 

This sets up the idea that multiple `rescue` clauses are more stylistic than a complex `if` or `case` statement wrapped up in a `if !result` statement. Also the `rescue` clauses are closer to the action that went wrong (being in the method that attempts the save) instead of being in the code that calls the saving method.

**Comment by Eneroth3** [comment]
Raising is the more proper way, no question about it. The one reason possible reason not to raise would be backward compatibility. However, if the status never can be false, even on failure, it should be somewhat safe to change the behavior.

**Comment by thomthom** [comment]
It's a good chance the return value doesn't change. I've come across a number of methods that will never return anything than `true`.

**Comment by thomthom** [comment]
I renamed this one to "Better failure handling for ComponentDefininition#save_as".

**Comment by DanRathbun** [comment]
I realized with a quick test today (using SU 2021.1.1,) that this method allows an empty string as an argument !

This results in a file named `".skp"` saved in the current working directory (which is usually the user's `"Documents"` path.)

This I would consider a bug. (Please add label and log Jira issue.)

**Comment by thomthom** [comment]
Is the summary of this:

* Make the method return `false` upon failure
* No allow empty filename

?

**Comment by DanRathbun** [comment]
> * No allow empty filename

Yes. It makes no good sense that a file named `".skp"` should be saved.

 Ie, the case of an empty filename string is itself likely due to a coding bug where the coder does not handle user input or user cancelling input correctly.

> * Make the method return `false` upon failure

We make the case above that raising **proper exceptions** is what this method (and several other IO methods in the API) should do. Ie...

* `ArgumentError` if the filepath arg is an **empty** string
* `Errno::EACCES` (permission denied) if permission to a path was denied
* `Errno::ENOENT` (no directory entry) if the directory path does not exist
* `Errno::EISDIR` (is a directory) if the string expands to an existing directory object.
 Ie, the code did not pass in a filename at the end of path string.
* `Errno::ENODEV` (no such device) if the drive specified does not exist.
* `Errno::ENOSPC` (no space on device) if the drive is full
* `Errno::EBUSY` (Device or resource busy) latency issue with network path.
* _etc._ ...

I am wondering if there can be general purpose "under the hood" write and read file functions for multiple API methods that raise these errors (or tell the using API method to raise these errors) ? This way proper IO handling can be maintained in one place in the code.

The alternative is to raise an [**`IOError`**](https://ruby-doc.org/core-2.7.2/IOError.html) for any such error (except the empty path string) and pass along the error message. 
Coders would then be required to test the error message string to know what happened. 
Elsewhere here it was argued (by JUlia I think,) that handling exceptions by comparing error messages is considered fragile and therefore poor practice.

**Comment by thomthom** [comment]
> We make the case above that raising proper exceptions is what this method (and several other IO methods in the API) should do.

With compatibility in mind, reducing risk of breaking existing extensions, what are your thoughts on this? We can't make changes to released API without considering impact. Ideal "should have" behaviour have to yield to minimize breakage.

**Comment by DanRathbun** [comment]
> With compatibility in mind, reducing risk of breaking existing extensions, what are your thoughts on this?

It is likely that coders never relied upon the return value. But yes, exceptions can stop code flow. 

So perhaps add an optional named `:raise_errors` boolean argument ?
```ruby
Sketchup.active_model.definitions[0].method(:save_as).arity
#=> 1
```
**That won't work for older versions.**

I was thinking ...
If the arg is `false` (the default) then simply `false` is returned upon error.
If it is `true`, then exceptions are raised.
This would give control to the coder.

But we'd need a new "save_as_with_status" method to give the coder control to raise exceptions.

**So yes I suppose returning `false` is all you can do for now.**

**Comment by Eneroth3** [comment]
I think just adding the return value without any new argument is good. Developers 10 years from now shouldn't have to copy in some boilerplate arity check to have the method work in a reasonable way.

These failures are rare and I don't expect extension developers to write custom error messages for 7 or so different errors. As long as the user gets an error message at all saying the save failed that is infinitely better than pretending it succeeded, and then the user can for instance often see within seconds that the file was already open.

**Comment by DanRathbun** [comment]
> I think just adding the return value without any new argument is good.

Which return value ?

I was just thinking during breakfast that since it had always returned `true`, and everything in Ruby is "truthy" except `nil` and `false` ... perhaps have it return integer error codes. Constants could be defined to give the codes better meaning.

> These failures are rare ...

I disagree. I've participated in many File IO error topics over the years. The top of the list would be network path / latency and permissions errors.

And I've lost count of the hundreds of error messages I've had to write myself.

> I don't expect extension developers to write custom error messages for 7 or so different errors.

They should not need to! The API (or a community library) should provide either a strings file for all distro'd languages or an error  handler module to get messages in the user's language (with args to insert filenames, etc.) This is something we had talked about many years ago, but has never happened.

So devs have had to do it all themselves duplicating effort.

----

But I feel I'm "beating the proverbial 'dead horse'". We'll not likely see anything more than `false` returned upon error.

**Comment by DanRathbun** [comment]
**Again**, a new developer needing better error support:

* https://forums.sketchup.com/t/definition-save-as-has-issue/202649
